File: Leetcode/Python/98. Validate Binary Search Tree/solution.py
# https://leetcode.com/problems/validate-binary-search-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        prev, bst_list = None, []

        # Inorder
        while root or bst_list:
            while root:
                bst_list.append(root)
                root = root.left
            root = bst_list.pop()

            if prev is not None and prev >= root.val:
                return False

            prev = root.val
            root = root.right

        return True


class Solution:  # Recursive
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.search(root, None, None)

    def search(self, root: Optional[TreeNode], min_value: int, max_value: int) -> bool:
        if not root:
            return True

        if (min_value is not None and root.val <= min_value) or (max_value is not None and max_value <= root.val):

            return False

        return self.search(root.left, min_value, root.val) and self.search(root.right, root.val, max_value)

File: Leetcode/Typescript/98. Validate Binary Search Tree/solution.ts
// https://leetcode.com/problems/validate-binary-search-tree/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
  let prev = null;
  const bstArray = [];

  // Inorder
  while (root !== null || bstArray.length > 0) {
    while (root !== null) {
      bstArray.push(root);
      root = root.left;
    }

    root = bstArray.pop();
    if (prev !== null && prev >= root.val) {
      return false;
    }

    prev = root.val;
    root = root.right;
  }
  return true;
}

// Recursive
function isValidBST(root: TreeNode | null): boolean {
  return rangeSearch(root, null, null);
}

function rangeSearch(
  root: TreeNode | null,
  minValue: number | null,
  maxValue: number | null
) {
  if (!root) {
    return true;
  }

  if (
    (minValue !== null && root.val <= minValue) ||
    (maxValue !== null && maxValue <= root.val)
  ) {
    return false;
  }

  const left = rangeSearch(root.left, minValue, root.val);
  if (!left) {
    return false;
  }

  const right = rangeSearch(root.right, root.val, maxValue);
  if (!right) {
    return false;
  }

  return true;
}
